### PR TITLE
Generalize exceptions to events

### PR DESCRIPTION
**Pull request by KarlSchimpf**
In order to make the exception handling proposal able to handle future events, the concept of an exception is extended to be an event. This PR introduces this concept.

### CODE REVIEW

File: proposals/Level-1.md [hunk 1]
```diff
@@ -51,13 +51,12 @@ exceptions.  For embedders that have garbage collection (such as JavaScript),
 this is not a problem.
 
 However, not all embedders may have a garbage collector. For this reason,
-WebAssembly exceptions are designed to allow the use of reference counters (to
-exceptions) to perform the the garbage collection in the embedder.
+WebAssembly exceptions are designed to allow the use of reference counters to
+perform the the garbage collection in the embedder.
```
Comment (eholk) [inline proposals/Level-1.md:55]: nit: 'the the' -> 'the'

I might rephrase this as "WebAssembly exceptions are designed to allow other storage management methods, such as reference counting, to perform the garbage collection in the embedder." I don't know what other methods their might be, but the idea is tracing GC is not required, using reference counting is an existence proof of this fact, but other methods are allowed too. We don't make any requirements on how the embedder manages storage, we simply say that it's possible to be done.

File: proposals/Level-1.md [hunk 2]
```diff
@@ -51,13 +51,12 @@ exceptions.  For embedders that have garbage collection (such as JavaScript),
...
 
-To do this, WebAssembly exceptions are immutable once created. This avoids
-cyclic data structures that can't be garbage collected. It also means that
-exceptions can't be stored into WebAssembly's linear memory. The rationale for
-this is twofold:
+To do this, WebAssembly exceptions are immutable once created, to avoid cyclic
+data structures that can't be garbage collected. It also means that exceptions
```
Comment (eholk) [inline proposals/Level-1.md:58]: nit: for spec text, I think spelling out contractions is better. i.e. "can't" -> "cannot"

File: proposals/Level-1.md [hunk 3]
```diff
@@ -78,94 +77,69 @@ instruction. Thrown exceptions are handled as follows:
 1. Throws not caught within a function body continue up the call stack, popping
    call frames, until an enclosing try block is found.
    
-1. If the call stack is exhausted without any enclosing try blocks, embedder
+1. If the call stack is exhausted without any enclosing try blocks, the embedder
    defines how to handle the uncaught exception.
    
 ### Event handling
```
Comment (eholk) [inline proposals/Level-1.md:83]: We talked about this offline some, but I'll mention it here in case other people have suggestions: I'm not sure "event" is the right term here, since it evokes the idea of events in the `addEventHandler` sense. I suggested something like "condition," "effect," or maybe "signal" but I'm not convinced these are better terms.

File: proposals/Level-1.md [hunk 4]
```diff
@@ -78,94 +77,69 @@ instruction. Thrown exceptions are handled as follows:
...
+general in its specification such that more general kinds of event handling can
+be added to WebAssembly in the future without having to redo exception handling.
 
-Events are generated by event yielding instructions. They yeild to an
-appropriate event handler. The event handler is then run. That event handler
-can either resume control at the point of the event handler, or optionally
-resume by sending values back to the yielding instruction allow the
-originating code to resume.
+An event handler allows one to process an event generated by a block of
+code. Events yield the current execution and look for a corresponding event
```
Comment (eholk) [inline proposals/Level-1.md:90]: What about "suspend" instead of "yield?"

File: proposals/Level-1.md [hunk 5]
```diff
@@ -78,94 +77,69 @@ instruction. Thrown exceptions are handled as follows:
...
+handler. If found, the corresponding event handler is run.  The event handler
+can optionally resume by sending values back to the yielded instruction,
+allowing the originating code to resume.
```
Comment (eholk) [inline proposals/Level-1.md:93]: Wait, we're not specing resumable exceptions yet, are we?

File: proposals/Level-1.md [hunk 6]
```diff
@@ -78,94 +77,69 @@ instruction. Thrown exceptions are handled as follows:
...
 
 Exceptions are a special case of an event in that they never resume. Similarly,
-a throw instruction is the yielding instruction of an exception, and the
-corresponding catching block is the event handler of the exception.
+a throw instruction is the yielding event of an exception. The catch block
+associated with a try block defines how to handle the throw.
```
Comment (eholk) [inline proposals/Level-1.md:97]: Is this implying that other kinds of events will have different instructions that go with them?

File: proposals/Level-1.md [hunk 7]
```diff
@@ -78,94 +77,69 @@ instruction. Thrown exceptions are handled as follows:
...
+section of a WebAssembly module. The event section is a list of event types
+that are defined by the module.
 
-The event section defines a list of event types that may be handled by the
-module.  Each `event type` has (at a minimum) an `event specifier` and a `type
-signature`. The event specifier defines the kind of event while the type
-signature defines the list of values associated with the event.  Depending on
+Each `event type` has a `event specifier` and a `type signature`. Depending on
 the value of the event specifier, additional values may be specified in the
-event type.
```
Comment (eholk) [inline proposals/Level-1.md:109]: Did you mean to delete this line?

File: proposals/Level-1.md [hunk 8]
```diff
@@ -78,94 +77,69 @@ instruction. Thrown exceptions are handled as follows:
...
-event. For exceptions, that can't resume, the result type must always be type
-'void'.
-
-Note: Resumable events may not use the `throw` instruction. Rather, they may use
-a new (yet to be defined) `yeild` instruction that passes value(s) to its
-enclosing contexts. Such an instruction would allow the calling context to
-return values back, and resume with the instruction following the yeild
-instruction, based on the values returned.
+event type corresponds to an exception. Future extensions my add additional
+event specifiers.
```
Comment (eholk) [inline proposals/Level-1.md:109]: Is it better to be more vague about event specifiers and just say that event entries are required to have a 0 byte, which is reserved for future expansion?

File: proposals/Level-1.md [hunk 9]
```diff
@@ -78,94 +77,69 @@ instruction. Thrown exceptions are handled as follows:
...
-corresponding runtime tag is called an `event tag`.
-
-Event types are numbered numerically, starting with the import section, and then
-followed by event types in the event section.  Each event type within a module
-are associated with unique event tags.
+The type signature of an event type is an index to the corresponding function
+signature defined in the `type` section. The parameter of the function signature
+defines the list of values associated with the event. The return value(s) of the
+signature defines what is passed back to yeilding instruction so that execution
+may resume.
```
Comment (eholk) [inline proposals/Level-1.md:115]: I think for now it'd be better to say that events cannot have a return value. I don't think we want resumable exceptions in the Level 1 proposal, even though we are trying to leave room for them.

File: proposals/Level-1.md [hunk 10]
```diff
@@ -78,94 +77,69 @@ instruction. Thrown exceptions are handled as follows:
...
 
-Event types can also be exported by adding appropriate descriptors in the export
-section of the module. Each imported/exported event index must be named to
-reconcile the corresponding (runtime) event tag.
+Within a module, event types are identified by an `event index` to the [event
+index space](#event-index-space). This (module specific) index identifies the
```
Comment (eholk) [inline proposals/Level-1.md:118]: nit: "This module-specific index"

File: proposals/Level-1.md [hunk 11]
```diff
@@ -78,94 +77,69 @@ instruction. Thrown exceptions are handled as follows:
...
 
-1. The `throw` instruction which creates a WebAssembly exception with the
-   corresponding exception tag, and then throws it.
+1. The `throw` instruction which creates a WebAssembly exception
+   with the corresponding exception tag, and then throws it.
 
-2. The `if_except` instruction that queries an exception to see if the exception
-   tag corresponding to the module's exception index. If true it pushes the
-   corresponding values of the exception onto the stack.
+2. The `if_except` instruction queries a exception to see if the corresponding
```
Comment (eholk) [inline proposals/Level-1.md:135]: nit: "an exception"

File: proposals/Level-1.md [hunk 12]
```diff
@@ -78,94 +77,69 @@ instruction. Thrown exceptions are handled as follows:
...
+   exception tag of the exception matches the wanted exception index.  If true
```
Comment (eholk) [inline proposals/Level-1.md:136]: "matches the _tag of_ the wanted exception index."

Just to be explicit that we are comparing tags to tags, and not tags to indices.

File: proposals/Level-1.md [hunk 13]
```diff
@@ -203,80 +176,93 @@ corresponding exception.
...
+The `if_except` block begins with an `if_except` instruction, and
+has two instruction blocks, defined by the `then` and `else` instructions
+like that of an `if` block.
+
+The `if_except` instruction queries of the exception on top of the stack, and
+then pops the exception. If the exception has the same exception tag as defined
+by the `if_except` instruction, the `then` block is entered. Otherwise the
+`else` block is entered. If the if_except block doesn't have an else block,
+control passes to the instruction following that block.
+The exception value is not popped when queried.
```
Comment (eholk) [inline proposals/Level-1.md:233]: In line 229 it says it pops the exception, but in line 233 it says the exception is not popped. In #44 I thought we decided to always pop the exception.

File: proposals/Level-1.md [hunk 14]
```diff
@@ -203,80 +176,93 @@ corresponding exception.
...
+
+If the query fails, it either enters the else block, or transfer control to the
+end of the if_except block if there is no else block.
 
 ### Stack traces
 
-When an exception is thrown, conceptually a runtime stack is associated with the
-exception. However, WebAssembly does not have (direct) access to the stack
-trace. Hence, it is left to the embedder to maintain and use at it sees fit.
+Earlier discussion implied that when an exception is thrown, the runtime will
```
Comment (eholk) [inline proposals/Level-1.md:261]: You can probably drop the reference to earlier discussion. I feel like this document should capture the current state of the proposal rather than the history.

File: proposals/Level-1.md [hunk 15]
```diff
@@ -300,9 +287,9 @@ instructions are *structured* control flow instructions, and can be
 labeled. This allows branch instructions to exit try and `if_except` blocks.
 
 The `except_index` of the `throw` and `if_except` instructions defines the
-exception type (and hence, exception tag) to create/extract form. See [event
-index space](#event-index-space) for further clarification of exception
-tags.
+exception value (and hence, exception tag) to create/extract form. See
```
Comment (eholk) [inline proposals/Level-1.md:290]: nit: form -> from

File: proposals/Level-1.md [hunk 16]
```diff
@@ -313,36 +300,38 @@ This section describes change in the
...
-import and event sections. Thus, the index space starts at zero with imported
-events followed by internally-defined events in the [event
+import and event sections. Thus, the index space starts at zero with
+imported events, followed by internally-defined events in the [event
 section](#event-section).
 
-The event index space defines the (module) static version of runtine event
-tags. These (runtime) event tags are guaranteed to be unique over all loaded
+The event index space defines the (module) static version of runtine
+event tags. For event indicies that are not imported/exported, the
```
Comment (eholk) [inline proposals/Level-1.md:308]: nit: indicies -> indices

File: proposals/Level-1.md [hunk 17]
```diff
@@ -313,36 +300,38 @@ This section describes change in the
...
+unique name imported/exported, and are aliased to the corresponding event index
+defined in each module.
 
 ## Changes to the binary model
 
-This section describes changes in the [binary encoding design
-document](https://github.com/WebAssembly/design/blob/master/BinaryEncoding.md).
+This section describes changes in
+the
+[binary encoding design document](https://github.com/WebAssembly/design/blob/master/BinaryEncoding.md).
```
Comment (eholk) [inline proposals/Level-1.md:320]: We can probably just reference the spec now that we have one, instead of the design doc.

File: proposals/Level-1.md [hunk 18]
```diff
@@ -313,36 +300,38 @@ This section describes change in the
...
-WebAssembly.
+An exception reference points to an exception. The size is defined by
+the embedder.
 
 ### Language Types
 
 | Opcode | Type constructor |
 |--------|------------------|
-| -0x0x5 |  `except_ref`    |
+| -0x05  |  `except_ref`    |
```
Comment (eholk) [inline proposals/Level-1.md:334]: I forget, did we decide to keep the type here, or move it up to 0x7whatever with `i32` and friends?

File: proposals/Level-1.md [hunk 19]
```diff
@@ -78,94 +77,69 @@ instruction. Thrown exceptions are handled as follows:
...
-signature` of the exception is the corresponding type signature of the event
-type. That is, the parameter list of the type signature defines the list of
-values associated with the exception, and the result type must be type `void`.
-
-Exceptions are identified by the by the corresponding event index, and is called
-an `exception index`. The corresponding (runtime) event tag is called the
-`exception tag`.  Like all event types, each exception type within a module
-(i.e. in the event and import sections) are associated with unique exception
-tags.
+An `exception` is an internal construct in WebAssembly that is maintained by the
```
Comment (rossberg) [inline proposals/Level-1.md:123]: Same comment here about "maintained by embedder". Perhaps avoid duplicating this text.

File: proposals/Level-1.md (hunk 12 above)
Comment (rossberg) [reply proposals/Level-1.md:136]: "matches the tag denoted by the exception index"

File: proposals/Level-1.md [hunk 20]
```diff
@@ -203,80 +176,93 @@ corresponding exception.
...
+end of the if_except block if there is no else block.
 
 ### Stack traces
 
-When an exception is thrown, conceptually a runtime stack is associated with the
-exception. However, WebAssembly does not have (direct) access to the stack
-trace. Hence, it is left to the embedder to maintain and use at it sees fit.
+Earlier discussion implied that when an exception is thrown, the runtime will
+pop the operand stack across function calls until a corresponding, enclosing try
+block is found. It also associates a stack trace that can be used to report
```
Comment (rossberg) [inline proposals/Level-1.md:263]: Nit: "It also associates" -> "It may also associate", since this is not really a Wasm requirement because it's not observable in Wasm.

File: proposals/Level-1.md [hunk 21]
```diff
@@ -313,36 +300,38 @@ This section describes change in the
...
-tags. These (runtime) event tags are guaranteed to be unique over all loaded
+The event index space defines the (module) static version of runtine
+event tags. For event indicies that are not imported/exported, the
+corresponding event tag is guaranteed to be unique over all loaded
 modules.
 
-For event indices imported/exported, unique event tags are created for
-each unique name imported/exported, and are aliased to the corresponding
-event index defined in each module.
+For event indices imported/exported, unique event tags are created for each
```
Comment (rossberg) [inline proposals/Level-1.md:312]: How about simply saying "Events that are imported or exported alias the respective events defined elsewhere, denoting the same tag."

File: proposals/Level-1.md [hunk 22]
```diff
@@ -313,36 +300,38 @@ This section describes change in the
...
 
 
 ### Data Types
 
 #### except_ref
 
-An exception reference points to an exception. The size is fixed, but unknown in
-WebAssembly.
+An exception reference points to an exception. The size is defined by
+the embedder.
```
Comment (rossberg) [inline proposals/Level-1.md:328]: I would drop this -- why require that there is a fixed size? That's not observable. In particular, I would avoid formulations that suggest that the semantics depends the an embedder.

File: proposals/Level-1.md (hunk 18 above)
Comment (rossberg) [reply proposals/Level-1.md:334]: It is a reference type, so it should be in the same cluster as `anyfunc`. Probably want to leave 0x41 for anyref, though.

File: proposals/Level-1.md [hunk 23]
```diff
@@ -389,55 +374,58 @@ or defined:
...
-| entries | `event_type* | list of defined event types.              |
 
 ##### Import section
 
-The import section is extended to include exception definitions by extending an
+The import section is extended to include event definitions by extending an
 `import_entry` as follows:
 
-If the `kind` is `event`:
+If the `kind` is `Event`:
```
Comment (rossberg) [inline proposals/Level-1.md:396]: Is this defined anywhere?

File: proposals/Level-1.md (hunk 3 above)
Comment (rossberg) [reply proposals/Level-1.md:83]: Btw, I'm a bit confused about the diff here. It seems like an earlier version of that section has already landed?

File: proposals/Level-1.md (hunk 18 above)
Comment (rossberg) [reply proposals/Level-1.md:334]: Sorry, I meant 0x11 when I said 0x41.

File: proposals/Level-1.md [hunk 24]
```diff
@@ -78,94 +77,69 @@ instruction. Thrown exceptions are handled as follows:
...
-corresponding catching block is the event handler of the exception.
+a throw instruction is the yielding event of an exception. The catch block
+associated with a try block defines how to handle the throw.
 
-An `event` is an internal construct in WebAssembly that is maintained by the
-embedder. WebAssembly events are defined by a new `event section` of a
-WebAssembly module.
+An `event` (such as an exception) is an internal construct in WebAssembly that
+is maintained by embedder. WebAssembly events are defined by a new `event`
+section of a WebAssembly module. The event section is a list of event types
```
[the remark this answers is not in the document]
Comment (rossberg) [reply proposals/Level-1.md:101]: Here is how I would put it: A tag is a value to distinguish different events, and it has a type that defines what values are attached to it when creating an event value with it. The event section is a list of definitions of event tags that are generated at runtime/inittime. An event index is a numeric name to refer to an (imported or defined) event tag within a module.

In practice, you can probably often say "event" as a short for "event tag" where it is clear.

File: proposals/Level-1.md (hunk 22 above)
Comment (rossberg) [reply proposals/Level-1.md:328]: Well, the specification does not enforce specific implementation strategies. A hypothetical engine could heap-allocate everything, eschew the stack entirely for CPS, and still be conforming.

File: proposals/Level-1.md (hunk 18 above)
Comment (rossberg) [reply proposals/Level-1.md:334]: My latest discussion with Luke seems to suggest that we gonna have a few more reference types in the ref proposal, so perhaps go for 0x18. :)

File: proposals/Level-1.md (hunk 23 above)
Comment (rossberg) [reply proposals/Level-1.md:396]: Ah sorry, didn't show up in the diff.
